### DISCUSSION TITLE
Understanding, plotting, and flipping 2D cell normals

### DISCUSSION THREAD

**Discussion by WadeRobertEvans** [pyvista/pyvista#1674] [Q&A] [answered]
Hello,

Apologies, I first put this query up in pyvista-support, but I then saw that it looks like we're now supposed to use this "Discussions" forum for queries. Please let me know if I should close / cancel the query in pyvista-support?

I have imported point cloud data (xyz coordinates) and performed triangulation using delaunay 2d to get a mesh of triangular faces / elements.

```py
  pdata = pv.PolyData(xyz_points)
  surf = pdata.delaunay_2d().clean()
```

I extracted the cell center points using:

```py
  centers = surf.cell_centers()
  center_pts = centers.points
```

I computed the normals using:
```
  surf_with_norms = surf.compute_normals(cell_normals=True,point_normals=False)
```

I plot the normals using:

```py
  pl = pv.Plotter()
  pl.add_mesh(surf_with_norms, show_edges=True, line_width=1)
  pl.add_axes()
  pl.background_color='w'
  pl.add_arrows(center_pts,surf_with_norms['Normals'],mag=20,color='red')
  pl.add_mesh(centers, color="g", point_size=10.0, render_points_as_spheres=True)
  pl.show()
```

Basically, I got arrays for the cell center points and the cell normals.
I used the cell center points array for the tail of the "normals arrow", and the cell normals array as the head of the "normals arrow" (refer to `pl.add_arrows(center_pts,surf_with_norms['Normals'],mag=20,color='red'))`.

Is this a correct method to plot the cell normals? I'm unsure of what "compute_normals" outputs, is it the head of the normal and is the tail of the normal supposed to be the cell center point?

For the cell normals array, I would've expected two coordinate points per cell normal (one for the tail and one for the head), but this doesn't seem to be the case when inspecting the array... There are the same number of cells as there are normals, both with only three columns (i.e. they have the same shape).

Below is a snippet of a plot showing the face centers and face normals.

![image](https://user-images.githubusercontent.com/84445181/133666087-7742b2a6-b116-452e-86cc-4491375daa39.png)

I will also need to flip the face normals.
Using `surf_with_norms.flip_normals()` doesn't seem to work.
Does this only flip the point normals?

In the end, I need the normals data to compare the data to that of ideal geometry (i.e. find the difference between the normals of the manufactured part to the normals of the ideal part).

Please assist!

**Comment by akaszynski** [comment, marked as answer]
> Please let me know if I should close / cancel the query in pyvista-support?

Yes, we're in the process of transferring it over and evaluating if this is better.  Let me know what you think.

---

### Normals flipping example

Here's how I would flip and plot normals:

```py
"""Demonstrate normal flipping"""
import pyvista as pv
pv.set_plot_theme('document')

# create a sample dataset
plane = pv.Plane(i_resolution=2, j_resolution=2).triangulate()

# copy and flip as flipping is inplace
plane_flipped = plane.copy()
plane_flipped.flip_normals()


pl = pv.Plotter(shape=(1, 2))

# on the left, plot the plane without flipping normals
pl.add_mesh(plane, style='wireframe', line_width=3)
pl.add_points(plane.cell_centers().points, point_size=15, color='g',
              render_points_as_spheres=True)
pl.add_arrows(plane.cell_centers().points,
              plane.cell_normals,
              mag=0.3,
              color='red')

# on the right, plot the plane after flipping normals
pl.subplot(0, 1)
pl.add_mesh(plane_flipped, style='wireframe', line_width=3)
pl.add_points(plane_flipped.cell_centers().points, point_size=15, color='g',
              render_points_as_spheres=True)
pl.add_arrows(plane_flipped.cell_centers().points,
              plane_flipped.cell_normals,
              mag=0.3,
              color='red')

# show both
pl.show()
```
![tmp](https://user-images.githubusercontent.com/11981631/133676902-3197e73d-89d5-4b5e-9435-958c379c866d.png)


> I will also need to flip the face normals.
Using surf_with_norms.flip_normals() doesn't seem to work.
Does this only flip the point normals?

[flip_normals](https://docs.pyvista.org/api/core/_autosummary/pyvista.PolyData.flip_normals.html?highlight=flip_normals#pyvista.PolyData.flip_normals) reorders the ordering of the faces and as a result flips cell and point normals.  I didn't play around with your code to see what the issue is; please try this example and let me know if it it solves your problem.

**Reply by WadeRobertEvans** [reply]
Thanks Alex! Your code works and it is cleaner and simpler. I see that there is no need to use compute_normals.

I closed the issue in pyvista-support.
